File: go/day5.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

type IntSet = map[int]bool

func setIntersection(set1, set2 IntSet) IntSet {
	intersection := make(IntSet)
	for num1 := range set1 {
		if _, ok := set2[num1]; ok {
			intersection[num1] = true
		}
	}

	return intersection
}

func parseInput(fname string) (incoming, outgoing map[int]IntSet, updates [][]int) {
	lines, err := ReadLines(fname)
	if err != nil {
		panic("Could not read lines")
	}

	incoming = make(map[int]IntSet)
	outgoing = make(map[int]IntSet)

	parseUpdates := false
	for _, line := range lines {
		if len(line) == 0 {
			parseUpdates = true
			continue
		}

		if parseUpdates {
			parts := strings.Split(line, ",")
			update := make([]int, len(parts))
			for i, part := range parts {
				num, _ := strconv.Atoi(part)
				update[i] = num
			}
			updates = append(updates, update)
		} else {
			parts := strings.Split(line, "|")
			num1, _ := strconv.Atoi(parts[0])
			num2, _ := strconv.Atoi(parts[1])

			if _, ok := outgoing[num1]; !ok {
				outgoing[num1] = make(IntSet)
			}
			if _, ok := incoming[num2]; !ok {
				incoming[num2] = make(IntSet)
			}
			outgoing[num1][num2] = true
			incoming[num2][num1] = true
		}
	}

	return
}

func topologicallyOrder(list []int, incoming, outgoing map[int]IntSet) []int {
	listSet := make(IntSet)
	for _, num := range list {
		listSet[num] = true
	}
	restrictedOutgoing := make(map[int]IntSet)
	restrictedIncoming := make(map[int]IntSet)
	for num, numOutgoing := range outgoing {
		restrictedOutgoing[num] = setIntersection(numOutgoing, listSet)
	}
	for num, numIncoming := range incoming {
		restrictedIncoming[num] = setIntersection(numIncoming, listSet)
	}

	empty := make([]int, 0)
	for _, num := range list {
		if len(restrictedIncoming[num]) == 0 {
			empty = append(empty, num)
		}
	}

	ordered := make([]int, 0, len(list))
	for len(empty) > 0 {
		num := empty[0]
		ordered = append(ordered, empty[0])
		empty = empty[1:]

		for neighbor := range restrictedOutgoing[num] {
			delete(restrictedIncoming[neighbor], num)
			if len(restrictedIncoming[neighbor]) == 0 {
				empty = append(empty, neighbor)
			}
		}
	}

	return ordered
}

func Day5a() {
	_, outgoing, updates := parseInput("../inputs/day5.txt")

	sum_good := 0
	for _, update := range updates {
		seen := make(IntSet)
		good := true
		for _, num := range update {
			intersection := setIntersection(seen, outgoing[num])
			if len(intersection) != 0 {
				good = false
				break
			}
			seen[num] = true
		}
		if good {
			sum_good += update[len(update)/2]
		}
	}

	fmt.Printf("%v\n", sum_good)
}

func Day5b() {
	incoming, outgoing, updates := parseInput("../inputs/day5.txt")

	sum_bad := 0
	for _, update := range updates {
		seen := make(IntSet)
		for _, num := range update {
			intersection := setIntersection(seen, outgoing[num])
			if len(intersection) != 0 {
				ordered := topologicallyOrder(update, incoming, outgoing)
				sum_bad += ordered[len(ordered)/2]
				break
			}
			seen[num] = true
		}
	}

	fmt.Printf("%v\n", sum_bad)
}

File: python/day5.py
import graphlib
from collections import defaultdict


def read_input(
    fname: str,
) -> tuple[dict[int, set[int]], list[list[int]]]:
    outgoing = defaultdict(set)
    updates = []
    with open(fname, "r") as file:
        updates_part = False
        for line in file:
            if line == "\n":
                # Start reading the "updates" part after the empty line
                updates_part = True
                continue

            if updates_part:
                parts = line.strip().split(",")
                updates.append([int(x) for x in parts])
            else:
                parts = line.strip().split("|")
                outgoing[int(parts[0])].add(int(parts[1]))

    return outgoing, updates


def day5a() -> None:
    outgoing, updates = read_input("../inputs/day5.txt")

    sum_good = 0
    for update in updates:
        seen = set()
        good = True
        for num in update:
            if seen.intersection(outgoing[num]):
                good = False
                break
            seen.add(num)

        if good:
            sum_good += update[len(update) // 2]

    print(sum_good)


def test_bad(update: list[int], outgoing: dict[int, set[int]]) -> int:
    update_set = set(update)

    outgoing_update = dict()
    for num in update:
        outgoing_update[num] = outgoing[num].intersection(update_set)

    toposort = graphlib.TopologicalSorter(outgoing_update)
    try:
        order = list(toposort.static_order())
        return order[len(order) // 2]
    except graphlib.CycleError:
        return 0


def day5b() -> None:
    outgoing, updates = read_input("../inputs/day5.txt")

    sum_bad = 0
    for update in updates:
        seen = set()
        for num in update:
            if seen.intersection(outgoing[num]):
                sum_bad += test_bad(update, outgoing)
                break
            seen.add(num)

    print(sum_bad)


day5b()
